### PR TITLE
Add preview mode functionality.

### PR DESCRIPTION
**Pull request by MaximAlien**
### Description

PR adds the ability to preview destination and routes via `PreviewViewController`.

For example:

https://user-images.githubusercontent.com/1496498/194180739-15f8e831-c5f5-482f-aecb-70499112909b.mov

### CODE REVIEW

File: Sources/MapboxNavigation/BannerContainerView.swift [hunk 1]
```diff
@@ -1,16 +1,20 @@
 import UIKit
 
 // :nodoc:
-@objc(MBBannerContainerView)
-open class BannerContainerView: UIView {
+public class Banner {
```
Comment (MaximAlien) [inline Sources/MapboxNavigation/BannerContainerView.swift:4]: At this point `Banner` and `Preview` classes serve as namespaces.

File: Sources/MapboxNavigation/BannerContainerView.swift [hunk 2]
```diff
@@ -1,16 +1,20 @@
...
     
     // :nodoc:
-    public enum `Type` {
-        case top
-        case bottom
+    public enum Position {
+        case topLeading
+        case bottomLeading
```
Comment (MaximAlien) [inline Sources/MapboxNavigation/BannerContainerView.swift:9]: Will be potentially extended to support also `topTrailing` and `bottomTrailing` cases, when top and bottom banner containers are shown on the right side of the screen.

File: Sources/MapboxNavigation/BannerPreviewing.swift [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+// :nodoc:
+public struct PreviewBannerConfiguration {
+    
+    // :nodoc:
+    public var position: Banner.Position = .bottomLeading
+    
+    // :nodoc:
+    public var height: CGFloat? = nil
```
Comment (MaximAlien) [inline Sources/MapboxNavigation/BannerPreviewing.swift:16]: Every banner that is presented in `PreviewViewController` must conform to `BannerPreviewing` protocol and provide default configuration via `PreviewBannerConfiguration`. I'm not yet sure if `height` property will remain here though. It'd be good to have the ability to set [detents](https://developer.apple.com/documentation/uikit/uisheetpresentationcontroller/3801903-detents) instead of plain height of the banner that can only be extended/collapsed. In any case this additional functionality will be added in other PR.

File: Sources/MapboxNavigation/PreviewViewController.swift [hunk 4]
```diff
@@ -0,0 +1,654 @@
...
+    
+    var bottomBanners = Stack<BannerPreviewing>()
+    
+    // :nodoc:
+    public var topmostBottomBanner: BannerPreviewing? {
+        bottomBanners.peek()
+    }
+    
+    // :nodoc:
+    @discardableResult public func popBanner(_ position: Banner.Position, animated: Bool = true) -> BannerPreviewing? {
```
Comment (MaximAlien) [inline Sources/MapboxNavigation/PreviewViewController.swift:38]: Preview functionality is not tied to destination/routes preview only. It's also possible to push and pop your own banner that conforms to `BannerPreviewing` protocol. I hope that this will provide more generic interface that can be extended in future. For example:
- the ability to show time-constrained alerts in top/bottom banners that are related to incidents
- feedback menu
- settings
- end of route view etc

Clients of this API must implement `UIViewController` that conforms to `BannerPreviewing` protocol, they will be then injected into banner container view provided by `NavigationView` API.

File: Sources/MapboxNavigation/PreviewViewController.swift [hunk 5]
```diff
@@ -0,0 +1,654 @@
...
+    func setupFloatingButtons() {
+        cameraModeFloatingButton = FloatingButton.rounded(imageEdgeInsets: UIEdgeInsets(floatLiteral: 12.0)) as CameraModeFloatingButton
+        cameraModeFloatingButton.delegate = self
+        
+        navigationView.floatingButtons = [
+            cameraModeFloatingButton
+        ]
+        
+#if DEBUG
+        let debugFloatingButton = FloatingButton.rounded(image: .debugImage,
```
Comment (MaximAlien) [inline Sources/MapboxNavigation/PreviewViewController.swift:335]: Functionality will be added in different PR.

**Review by Udumft — APPROVED** [review]
LGTM with few questions

File: Sources/MapboxNavigation/BannerContainerView.swift [hunk 6]
```diff
@@ -1,16 +1,20 @@
 import UIKit
 
 // :nodoc:
-@objc(MBBannerContainerView)
-open class BannerContainerView: UIView {
+public class Banner {
     
     // :nodoc:
-    public enum `Type` {
```
Comment (1ec5) [inline Sources/MapboxNavigation/BannerContainerView.swift:8]: Removing this public enumeration would normally be a backwards-incompatible change. This class is undocumented, but it has been around for a while. I didn’t spot much usage of BannerContainerView [elsewhere on GitHub](https://github.com/search?q=BannerContainerView+mapboxnavigation&type=code), so I think it should be OK.

File: Sources/MapboxNavigation/PreviewViewController.swift [hunk 7]
```diff
@@ -0,0 +1,656 @@
...
+            }
+        }
+        
+        return navigationView
+    }
+    
+    func setupBackButton() {
+        let backButton = BackButton(type: .system)
+        backButton.translatesAutoresizingMaskIntoConstraints = false
+        // TODO: Add localization.
```
[the remark this answers is not in the document]
Comment (1ec5) [reply Sources/MapboxNavigation/PreviewViewController.swift:310]: Yes, we should make the “Back” string localizable at a minimum, if not in this PR then as tail work.

File: Sources/MapboxNavigation/PreviewViewController.swift [hunk 8]
```diff
@@ -0,0 +1,656 @@
...
+        backButton.setTitle("Back", for: .normal)
```
Comment (1ec5) [inline Sources/MapboxNavigation/PreviewViewController.swift:311]: While we’re at it, we should make the button title customizable. On iOS, a UINavigationController back button often inherits the name of the previous view controller on the stack when provided. I don’t know if there’s a straightforward way to do that automatically – maybe by crawling up the responder chain? But in that case, the button would need to be constrained to a maximum width and truncated, in case the previous view controller’s title is long.

File: Sources/MapboxNavigation/BannerContainerView.swift [hunk 9]
```diff
@@ -207,24 +211,25 @@ open class BannerContainerView: UIView {
...
             }
             
             expansionConstraint.isActive = true
         }
         
+        setNeedsLayout()
         layoutIfNeeded()
         
         if animated {
             // TODO: Improve animation for devices with notch.
```
Comment (1ec5) [inline Sources/MapboxNavigation/BannerContainerView.swift:228]: Remember to ticket out any to-do items, if you haven’t already.

File: Sources/MapboxNavigation/PreviewViewController.swift (hunk 5 above)
Comment (1ec5) [reply Sources/MapboxNavigation/PreviewViewController.swift:335]: Is this is a feature we’re adding long-term, or just a placeholder for a different floating button we’ll be adding by default? If the latter, remember to remove the debug icon from the asset catalog.

File: Sources/MapboxNavigation/Resources/Assets.xcassets/debug.imageset/Contents.json [hunk 10]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "debug.pdf",
```
Comment (1ec5) [inline Sources/MapboxNavigation/Resources/Assets.xcassets/debug.imageset/Contents.json:4]: Nice icon. If we intend to keep it around long-term, please add it to [navigation-ui-resources](https://github.com/mapbox/navigation-ui-resources/) (assuming it’s in the public domain).

File: Sources/MapboxNavigation/PreviewOptions.swift [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import MapboxCoreNavigation
+
+// :nodoc:
```
Comment (1ec5) [inline Sources/MapboxNavigation/PreviewOptions.swift:3]: Do we intend to keep all these new types undocumented once the feature fully lands? The main reason we used to sprinkle `:nodoc:` everywhere was that we needed to expose various UI classes to UIAppearance for styling but otherwise wouldn’t have wanted them to be public. Is that the case for these types too?

File: Sources/MapboxNavigation/CameraModeFloatingButton.swift [hunk 12]
```diff
@@ -62,7 +57,7 @@ class CameraModeFloatingButton: FloatingButton {
         case .centered:
             cameraMode = .following
         case .following:
-            cameraMode = .idle
+            cameraMode = .centered
```
[the remark this answers is not in the document]
Comment (MaximAlien) [reply Sources/MapboxNavigation/CameraModeFloatingButton.swift:81]: Yes, intention is to move map to idle state only after interaction with map itself.

File: Sources/MapboxNavigation/Preview.swift [hunk 13]
```diff
@@ -0,0 +1,11 @@
+import MapboxMaps
+
+class Preview {
```
[the remark this answers is not in the document]
Comment (MaximAlien) [reply Sources/MapboxNavigation/Preview.swift:3]: Serves role of a namespace, don't see any other good ways of doing this in Swift.

File: Sources/MapboxNavigation/PreviewViewController.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (MaximAlien) [reply Sources/MapboxNavigation/PreviewViewController.swift]: Added only English localization for now, will create a ticket so that we track other languages as well.

File: Sources/MapboxNavigation/PreviewOptions.swift (hunk 11 above)
Comment (MaximAlien) [reply Sources/MapboxNavigation/PreviewOptions.swift:3]: Most of the publicly exposed symbols will be documented in scope of different PR. I still plan to leave some public symbols undocumented because I'm not 100% sure whether they're really something that we'll need.

File: Sources/MapboxNavigation/PreviewViewController.swift (hunk 5 above)
Comment (MaximAlien) [reply Sources/MapboxNavigation/PreviewViewController.swift:148]: Planning on keeping it long-term, it'll be useful mostly for debugging purposes (showing viewport padding, changing runtime styling, logging etc).

File: Sources/MapboxNavigation/BannerPresentation.swift [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+                   completion: (() -> Void)?) -> Banner?
+    
+    func push(_ banner: Banner,
+              animated: Bool,
+              duration: TimeInterval,
+              animations: (() -> Void)?,
+              completion: (() -> Void)?)
+}
+
+extension BannerPresentation {
```
Comment (MaximAlien) [inline Sources/MapboxNavigation/BannerPresentation.swift:28]: I plan to use default implementation of `BannerPresentation` protocol in both `PreviewViewController` and `NavigationViewController`.
